// react-medium-image-zoom/dist/utils.js
function isElement(el) {
    if (typeof Element === 'undefined') {
        return false;
    }
    return el instanceof Element;
}
const testElType = (type, el) => isElement(el) && el.tagName.toUpperCase() === type;
export const testDiv = (el) => testElType('DIV', el) || testElType('SPAN', el);
export const testImg = (el) => testElType('IMG', el);
export const testImgLoaded = (el) => el.complete && el.naturalHeight !== 0;
export const testSvg = (el) => testElType('SVG', el);
const getScaleToWindow = ({ height, offset, width }) => Math.min((window.innerWidth - offset * 2) / width, (window.innerHeight - offset * 2) / height);
const getScaleToWindowMax = ({ containerHeight, containerWidth, offset, targetHeight, targetWidth, }) => {
    const scale = getScaleToWindow({
        height: targetHeight,
        offset,
        width: targetWidth,
    });
    const ratio = targetWidth > targetHeight
        ? targetWidth / containerWidth
        : targetHeight / containerHeight;
    return scale > 1 ? ratio : scale * ratio;
};
const getScale = ({ containerHeight, containerWidth, hasScalableSrc, offset, targetHeight, targetWidth, }) => {
    if (containerHeight === 0 || containerWidth === 0) {
        return 1;
    }
    return !hasScalableSrc && targetHeight !== 0 && targetWidth !== 0
        ? getScaleToWindowMax({
            containerHeight,
            containerWidth,
            offset,
            targetHeight,
            targetWidth,
        })
        : getScaleToWindow({
            height: containerHeight,
            offset,
            width: containerWidth,
        });
};
const URL_REGEX = /url(?:\(['"]?)(?<url>.*?)(?:['"]?\))/;
export const getImgSrc = imgEl => {
    if (imgEl !== null) {
        if (testImg(imgEl)) {
            return imgEl.currentSrc === '' ? undefined : imgEl.currentSrc;
        }
        else if (testDiv(imgEl)) {
            const { backgroundImage: bgImg } = window.getComputedStyle(imgEl);
            if (bgImg !== '') {
                return URL_REGEX.exec(bgImg)?.[1];
            }
        }
    }
};
export const getImgAlt = imgEl => {
    if (imgEl !== null) {
        if (testImg(imgEl)) {
            return imgEl.alt;
        }
        else {
            return imgEl.getAttribute('aria-label') ?? undefined;
        }
    }
};
const getImgRegularStyle = ({ containerHeight, containerLeft, containerTop, containerWidth, hasScalableSrc, offset, targetHeight, targetWidth, }) => {
    const scale = getScale({
        containerHeight,
        containerWidth,
        hasScalableSrc,
        offset,
        targetHeight,
        targetWidth,
    });
    return {
        top: containerTop,
        left: containerLeft,
        width: containerWidth * scale,
        height: containerHeight * scale,
        transform: `translate(0,0) scale(${1 / scale})`,
    };
};
const parsePosition = ({ position, relativeNum }) => {
    const positionNum = parseFloat(position);
    return position.endsWith('%')
        ? (relativeNum * positionNum) / 100
        : positionNum;
};
const getImgObjectFitStyle = ({ containerHeight, containerLeft, containerTop, containerWidth, hasScalableSrc, objectFit: objectFitParam, objectPosition, offset, targetHeight, targetWidth, }) => {
    let resolvedObjectFit = objectFitParam;
    if (resolvedObjectFit === 'scale-down') {
        if (targetWidth <= containerWidth && targetHeight <= containerHeight) {
            resolvedObjectFit = 'none';
        }
        else {
            resolvedObjectFit = 'contain';
        }
    }
    if (resolvedObjectFit === 'cover' || resolvedObjectFit === 'contain') {
        const widthRatio = containerWidth / targetWidth;
        const heightRatio = containerHeight / targetHeight;
        const ratio = resolvedObjectFit === 'cover'
            ? Math.max(widthRatio, heightRatio)
            : Math.min(widthRatio, heightRatio);
        const [posLeft = '50%', posTop = '50%'] = objectPosition.split(' ');
        const posX = parsePosition({
            position: posLeft,
            relativeNum: containerWidth - targetWidth * ratio,
        });
        const posY = parsePosition({
            position: posTop,
            relativeNum: containerHeight - targetHeight * ratio,
        });
        const scale = getScale({
            containerHeight: targetHeight * ratio,
            containerWidth: targetWidth * ratio,
            hasScalableSrc,
            offset,
            targetHeight,
            targetWidth,
        });
        return {
            top: containerTop + posY,
            left: containerLeft + posX,
            width: targetWidth * ratio * scale,
            height: targetHeight * ratio * scale,
            transform: `translate(0,0) scale(${1 / scale})`,
        };
    }
    else if (resolvedObjectFit === 'none') {
        const [posLeft = '50%', posTop = '50%'] = objectPosition.split(' ');
        const posX = parsePosition({
            position: posLeft,
            relativeNum: containerWidth - targetWidth,
        });
        const posY = parsePosition({
            position: posTop,
            relativeNum: containerHeight - targetHeight,
        });
        const scale = getScale({
            containerHeight: targetHeight,
            containerWidth: targetWidth,
            hasScalableSrc,
            offset,
            targetHeight,
            targetWidth,
        });
        return {
            top: containerTop + posY,
            left: containerLeft + posX,
            width: targetWidth * scale,
            height: targetHeight * scale,
            transform: `translate(0,0) scale(${1 / scale})`,
        };
    }
    else if (resolvedObjectFit === 'fill') {
        const widthRatio = containerWidth / targetWidth;
        const heightRatio = containerHeight / targetHeight;
        const ratio = Math.max(widthRatio, heightRatio);
        const scale = getScale({
            containerHeight: targetHeight * ratio,
            containerWidth: targetWidth * ratio,
            hasScalableSrc,
            offset,
            targetHeight,
            targetWidth,
        });
        return {
            width: containerWidth * scale,
            height: containerHeight * scale,
            transform: `translate(0,0) scale(${1 / scale})`,
        };
    }
    else {
        return {};
    }
};
const getDivImgStyle = ({ backgroundPosition, backgroundSize, containerHeight, containerLeft, containerTop, containerWidth, hasScalableSrc, offset, targetHeight, targetWidth, }) => {
    if (backgroundSize === 'cover' || backgroundSize === 'contain') {
        const widthRatio = containerWidth / targetWidth;
        const heightRatio = containerHeight / targetHeight;
        const ratio = backgroundSize === 'cover'
            ? Math.max(widthRatio, heightRatio)
            : Math.min(widthRatio, heightRatio);
        const [posLeft = '50%', posTop = '50%'] = backgroundPosition.split(' ');
        const posX = parsePosition({
            position: posLeft,
            relativeNum: containerWidth - targetWidth * ratio,
        });
        const posY = parsePosition({
            position: posTop,
            relativeNum: containerHeight - targetHeight * ratio,
        });
        const scale = getScale({
            containerHeight: targetHeight * ratio,
            containerWidth: targetWidth * ratio,
            hasScalableSrc,
            offset,
            targetHeight,
            targetWidth,
        });
        return {
            top: containerTop + posY,
            left: containerLeft + posX,
            width: targetWidth * ratio * scale,
            height: targetHeight * ratio * scale,
            transform: `translate(0,0) scale(${1 / scale})`,
        };
    }
    else if (backgroundSize === 'auto') {
        const [posLeft = '50%', posTop = '50%'] = backgroundPosition.split(' ');
        const posX = parsePosition({
            position: posLeft,
            relativeNum: containerWidth - targetWidth,
        });
        const posY = parsePosition({
            position: posTop,
            relativeNum: containerHeight - targetHeight,
        });
        const scale = getScale({
            containerHeight: targetHeight,
            containerWidth: targetWidth,
            hasScalableSrc,
            offset,
            targetHeight,
            targetWidth,
        });
        return {
            top: containerTop + posY,
            left: containerLeft + posX,
            width: targetWidth * scale,
            height: targetHeight * scale,
            transform: `translate(0,0) scale(${1 / scale})`,
        };
    }
    else {
        const [sizeW = '50%', sizeH = '50%'] = backgroundSize.split(' ');
        const sizeWidth = parsePosition({
            position: sizeW,
            relativeNum: containerWidth,
        });
        const sizeHeight = parsePosition({
            position: sizeH,
            relativeNum: containerHeight,
        });
        const widthRatio = sizeWidth / targetWidth;
        const heightRatio = sizeHeight / targetHeight;
        const ratio = Math.min(widthRatio, heightRatio);
        const [posLeft = '50%', posTop = '50%'] = backgroundPosition.split(' ');
        const posX = parsePosition({
            position: posLeft,
            relativeNum: containerWidth - targetWidth * ratio,
        });
        const posY = parsePosition({
            position: posTop,
            relativeNum: containerHeight - targetHeight * ratio,
        });
        const scale = getScale({
            containerHeight: targetHeight * ratio,
            containerWidth: targetWidth * ratio,
            hasScalableSrc,
            offset,
            targetHeight,
            targetWidth,
        });
        return {
            top: containerTop + posY,
            left: containerLeft + posX,
            width: targetWidth * ratio * scale,
            height: targetHeight * ratio * scale,
            transform: `translate(0,0) scale(${1 / scale})`,
        };
    }
};
const SRC_SVG_REGEX = /\.svg$/i;
export const getStyleModalImg = ({ hasZoomImg, imgSrc, isSvg, isZoomed, loadedImgEl, offset, shouldRefresh, targetEl, }) => {
    const hasScalableSrc = isSvg ||
        imgSrc?.slice(0, 18) === 'data:image/svg+xml' ||
        hasZoomImg ||
        (imgSrc !== undefined && SRC_SVG_REGEX.test(imgSrc));
    const imgRect = targetEl.getBoundingClientRect();
    const targetElComputedStyle = window.getComputedStyle(targetEl);
    const isDivImg = loadedImgEl != null && testDiv(targetEl);
    const isImgObjectFit = loadedImgEl != null && !isDivImg;
    const targetHeight = loadedImgEl != null && loadedImgEl.naturalHeight !== 0
        ? loadedImgEl.naturalHeight
        : imgRect.height;
    const targetWidth = loadedImgEl != null && loadedImgEl.naturalWidth !== 0
        ? loadedImgEl.naturalWidth
        : imgRect.width;
    const styleImgRegular = getImgRegularStyle({
        containerHeight: imgRect.height,
        containerLeft: imgRect.left,
        containerTop: imgRect.top,
        containerWidth: imgRect.width,
        hasScalableSrc,
        offset,
        targetHeight,
        targetWidth,
    });
    const styleImgObjectFit = isImgObjectFit
        ? getImgObjectFitStyle({
            containerHeight: imgRect.height,
            containerLeft: imgRect.left,
            containerTop: imgRect.top,
            containerWidth: imgRect.width,
            hasScalableSrc,
            objectFit: targetElComputedStyle.objectFit,
            objectPosition: targetElComputedStyle.objectPosition,
            offset,
            targetHeight,
            targetWidth,
        })
        : undefined;
    const styleDivImg = isDivImg
        ? getDivImgStyle({
            backgroundPosition: targetElComputedStyle.backgroundPosition,
            backgroundSize: targetElComputedStyle.backgroundSize,
            containerHeight: imgRect.height,
            containerLeft: imgRect.left,
            containerTop: imgRect.top,
            containerWidth: imgRect.width,
            hasScalableSrc,
            offset,
            targetHeight,
            targetWidth,
        })
        : undefined;
    const style = {
        ...styleImgRegular,
        ...styleImgObjectFit,
        ...styleDivImg,
    };
    if (isZoomed) {
        const viewportX = window.innerWidth / 2;
        const viewportY = window.innerHeight / 2;
        const childCenterX = parseFloat(String(style.left ?? 0)) +
            parseFloat(String(style.width ?? 0)) / 2;
        const childCenterY = parseFloat(String(style.top ?? 0)) +
            parseFloat(String(style.height ?? 0)) / 2;
        const translateX = viewportX - childCenterX;
        const translateY = viewportY - childCenterY;
        if (shouldRefresh) {
            style.transitionDuration = '0.01ms';
        }
        style.transform = `translate(${translateX}px,${translateY}px) scale(1)`;
    }
    return style;
};
export const getStyleGhost = imgEl => {
    if (imgEl == null) {
        return {};
    }
    if (testSvg(imgEl)) {
        const { parentElement: parentEl } = imgEl;
        const rect = imgEl.getBoundingClientRect();
        if (parentEl == null) {
            return {
                height: rect.height,
                left: rect.left,
                width: rect.width,
                top: rect.top,
            };
        }
        else {
            const parentRect = parentEl.getBoundingClientRect();
            return {
                height: rect.height,
                left: parentRect.left - rect.left,
                top: parentRect.top - rect.top,
                width: rect.width,
            };
        }
    }
    else {
        return {
            height: imgEl.offsetHeight,
            left: imgEl.offsetLeft,
            width: imgEl.offsetWidth,
            top: imgEl.offsetTop,
        };
    }
};
export const adjustSvgIDs = (svgEl) => {
    const newIdSuffix = '-zoom';
    const attrs = [
        'clip-path',
        'fill',
        'mask',
        'marker-start',
        'marker-mid',
        'marker-end',
    ];
    const idMap = new Map();
    if (svgEl.hasAttribute('id')) {
        const { id: oldId } = svgEl;
        const newId = oldId + newIdSuffix;
        idMap.set(oldId, newId);
        const svgNode = svgEl;
        svgNode.id = newId;
    }
    svgEl.querySelectorAll('[id]').forEach(el => {
        const { id: oldId } = el;
        const newId = oldId + newIdSuffix;
        idMap.set(oldId, newId);
        const node = el;
        node.id = newId;
    });
    idMap.forEach((newId, oldId) => {
        const urlOldID = `url(#${oldId})`;
        const urlNewID = `url(#${newId})`;
        const attrsQuery = attrs.map(attr => `[${attr}="${urlOldID}"]`).join(', ');
        svgEl.querySelectorAll(attrsQuery).forEach(usedEl => {
            attrs.forEach(attr => {
                if (usedEl.getAttribute(attr) === urlOldID) {
                    usedEl.setAttribute(attr, urlNewID);
                }
            });
        });
    });
    svgEl.querySelectorAll('style').forEach(styleEl => {
        idMap.forEach((newId, oldId) => {
            const { textContent } = styleEl;
            if (textContent !== '') {
                const styleNode = styleEl;
                styleNode.textContent = textContent.replaceAll(`#${oldId}`, `#${newId}`);
            }
        });
    });
};
